File: netlify/functions/_lib/usage-tracker.js
const fs = require("node:fs");
const path = require("node:path");
const { createLogger } = require("./logger");

const logger = createLogger("usage_tracker");

/**
 * Трекер использования токенов OpenAI API.
 *
 * Записывает статистику вызовов в usage.json:
 * - Общее количество вызовов
 * - Суммарное количество токенов
 * - Детальный лог каждого запроса с timestamp, моделью, токенами
 *
 * Для локальной разработки пишет в корень проекта.
 * В serverless окружении (Netlify) пытается писать в /tmp (эфемерное хранилище).
 * Ошибки записи не прерывают основной запрос (best-effort logging).
 */
class UsageTracker {
  constructor() {
    this.usageFile = this.resolveUsageFilePath();
    this.ensureUsageFile();
  }

  /**
   * Определяет доступный путь для записи usage.json.
   *
   * Пробует кандидатов в порядке приоритета:
   * 1. ./usage.json (локальная разработка)
   * 2. /tmp/getanalyst/usage.json (serverless окружение)
   *
   * @returns {string} Путь к файлу usage.json (может быть недоступен для записи)
   */
  resolveUsageFilePath() {
    const candidates = [path.resolve(process.cwd(), "usage.json"), path.resolve("/tmp/getanalyst/usage.json")];

    for (const candidate of candidates) {
      try {
        fs.mkdirSync(path.dirname(candidate), { recursive: true });
        fs.appendFileSync(candidate, "", "utf-8");
        return candidate;
      } catch (_error) {}
    }

    return candidates[0];
  }

  /**
   * Создаёт usage.json с начальной структурой, если файл не существует.
   *
   * @throws Не выбрасывает исключения - логирует предупреждение при ошибке
   */
  ensureUsageFile() {
    try {
      if (!fs.existsSync(this.usageFile) || fs.statSync(this.usageFile).size === 0) {
        fs.writeFileSync(
          this.usageFile,
          JSON.stringify({ calls: 0, total_tokens: 0, log: [] }, null, 2),
          "utf-8"
        );
      }
    } catch (error) {
      logger.warning(`Usage storage unavailable: ${error.message}`);
    }
  }

  /**
   * Читает текущую статистику из usage.json.
   *
   * @returns {Object} Объект статистики: {calls: number, total_tokens: number, log: Array}
   * @returns {number} return.calls - Общее количество вызовов API
   * @returns {number} return.total_tokens - Суммарное количество токенов
   * @returns {Array} return.log - Массив записей с деталями каждого вызова
   */
  readUsage() {
    try {
      return JSON.parse(fs.readFileSync(this.usageFile, "utf-8"));
    } catch {
      return { calls: 0, total_tokens: 0, log: [] };
    }
  }

  /**
   * Записывает обновлённую статистику в usage.json.
   *
   * @param {Object} data - Объект статистики для записи
   */
  writeUsage(data) {
    fs.writeFileSync(this.usageFile, JSON.stringify(data, null, 2), "utf-8");
  }

  /**
   * Фиксирует статистику вызова OpenAI API.
   *
   * Добавляет запись в лог и обновляет счётчики.
   * Ошибки записи намеренно не пробрасываются, чтобы не ломать основной ответ пользователю.
   *
   * @param {Object} params - Параметры вызова для трекинга
   * @param {string} params.model - ID модели OpenAI (например, "gpt-5.4")
   * @param {number} params.promptTokens - Количество токенов в промпте
   * @param {number} params.completionTokens - Количество токенов в ответе
   * @param {number|null} [params.temperature=null] - Температура генерации
   * @param {string|null} [params.validationReason=null] - Результат валидации промпта
   * @param {string|null} [params.validatorResponse=null] - Сырой ответ валидатора
   *
   * @example
   * tracker.track({
   *   model: "gpt-5.4",
   *   promptTokens: 312,
   *   completionTokens: 480,
   *   temperature: 0.7,
   *   validationReason: "ok"
   * });
   */
  track({
    model,
    promptTokens,
    completionTokens,
    temperature = null,
    validationReason = null,
    validatorResponse = null
  }) {
    try {
      const data = this.readUsage();
      const totalTokens = Number(promptTokens || 0) + Number(completionTokens || 0);
      data.calls += 1;
      data.total_tokens += totalTokens;

      const logEntry = {
        timestamp: new Date().toISOString(),
        model,
        prompt_tokens: promptTokens || 0,
        completion_tokens: completionTokens || 0,
        total_tokens: totalTokens
      };

      if (temperature !== null && temperature !== undefined) {
        logEntry.temperature = temperature;
      }
      if (validationReason !== null) {
        logEntry.validation_reason = validationReason;
      }
      if (validatorResponse !== null) {
        logEntry.validator_response = validatorResponse;
      }

      data.log.push(logEntry);
      this.writeUsage(data);
    } catch (error) {
      logger.error(`Failed to track usage: ${error.message}`);
    }
  }
}

module.exports = { UsageTracker };

File: python_legacy/backend/services/usage_tracker.py
"""
Usage tracking service for monitoring token consumption.

Tracks:
1. Total API calls
2. Token usage per call
3. Historical log of all requests
"""

import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional
from backend.logger import setup_logger

logger = setup_logger(__name__)


class UsageTracker:
    """Tracks and persists LLM API usage statistics."""

    def __init__(self, usage_file: str = "usage.json"):
        """
        Initialize usage tracker.

        Args:
            usage_file: Path to usage.json file (relative to project root)
        """
        self.usage_file = Path(usage_file)
        self._ensure_usage_file()

    def _ensure_usage_file(self) -> None:
        """Create usage.json if it doesn't exist or is empty."""
        if not self.usage_file.exists() or self.usage_file.stat().st_size == 0:
            initial_data = {
                "calls": 0,
                "total_tokens": 0,
                "log": []
            }
            self._write_usage(initial_data)
            logger.info(f"Created new usage file: {self.usage_file}")

    def _read_usage(self) -> dict:
        """
        Read current usage data from file.

        Returns:
            Dict with keys: calls, total_tokens, log
        """
        try:
            with open(self.usage_file, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception as e:
            logger.error(f"Failed to read usage file: {e}")
            # Return default structure if file is corrupted
            return {
                "calls": 0,
                "total_tokens": 0,
                "log": []
            }

    def _write_usage(self, data: dict) -> None:
        """
        Write usage data to file.

        Args:
            data: Usage data dict
        """
        try:
            with open(self.usage_file, "w", encoding="utf-8") as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
        except Exception as e:
            logger.error(f"Failed to write usage file: {e}")

    def track(
        self,
        model: str,
        prompt_tokens: int,
        completion_tokens: int,
        temperature: Optional[float] = None,
        validation_reason: Optional[str] = None,
        validator_response: Optional[str] = None
    ) -> None:
        """
        Record a new API call in usage.json.

        Args:
            model: Model ID used for the call
            prompt_tokens: Number of tokens in prompt
            completion_tokens: Number of tokens in completion
            temperature: Temperature parameter (if used)
            validation_reason: Validation result reason (ok, injection, off_topic)
            validator_response: Raw response from validator model
        """
        try:
            # Read current data
            data = self._read_usage()

            # Calculate total tokens for this call
            total_tokens = prompt_tokens + completion_tokens

            # Update counters
            data["calls"] += 1
            data["total_tokens"] += total_tokens

            # Create log entry
            log_entry = {
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "model": model,
                "prompt_tokens": prompt_tokens,
                "completion_tokens": completion_tokens,
                "total_tokens": total_tokens
            }

            # Add temperature if provided
            if temperature is not None:
                log_entry["temperature"] = temperature

            # Add validation info if provided
            if validation_reason is not None:
                log_entry["validation_reason"] = validation_reason
            if validator_response is not None:
                log_entry["validator_response"] = validator_response

            # Append to log
            data["log"].append(log_entry)

            # Write back to file
            self._write_usage(data)

            logger.info(
                f"Usage tracked: model={model}, "
                f"tokens={total_tokens} (prompt={prompt_tokens}, completion={completion_tokens}), "
                f"validation={validation_reason}"
            )

        except Exception as e:
            logger.error(f"Failed to track usage: {e}")
            # Don't raise - tracking failure shouldn't break the main flow

    def get_stats(self) -> dict:
        """
        Get current usage statistics.

        Returns:
            Dict with keys: calls, total_tokens, log
        """
        return self._read_usage()
